File: java/class-diagram/mitologico/MundoMitologico.java
// Autor: Luis Espino
// 2020

abstract class Caballo{
    String color;    
    Caballo(){
        this.color = "blanco";
    }
    void caminar(){};
    void correr(){};
    abstract void parar();
}

interface Alas{
    void volar();
}

interface Cuerno{
    void cornear();
}

interface Torso{
    void hablar();
}

class Pegaso extends Caballo implements Alas{
    String color;
    Pegaso(String color){
    	this.color = color;
    }
    void show(){
        System.out.println(color);
    }
    void parar() {};
    public void volar() {};
}

class Unicornio extends Caballo implements Cuerno{
    void parar(){};
    public void cornear(){};
    Unicornio(){
    	super();
    } 
    void show(){
    	System.out.println(color);
    }

}

class Centauro extends Caballo implements Torso{
    String color;
    void parar(){}
    public void hablar(){}
    Centauro(String color){
    	this.color = color;    	
    }
    void show()    {
    	System.out.println(color);
    }
}

public class MundoMitologico {
    Pegaso p = new Pegaso("negro");
    Unicornio u = new Unicornio();
    Centauro c = new Centauro("cafe");

    public static void main(String[] args) {
	MundoMitologico m = new MundoMitologico();
	m.p.show();
	m.u.show();
	m.c.show();
    }
}

File: 02-object-oriented/cpp/mitologico/MundoMitologico.cpp
// Luis Espino 2020

#include <iostream>

using namespace std;

// clase abstracta
class Caballo
{
    void caminar(){}
    void correr(){}
    virtual void parar() = 0;
protected:
    string color;
public:
    Caballo(){
        color = "blanco";
    }
};

// interfaces
class Alas
{
    virtual void volar() = 0;
};

class Cuerno
{
    virtual void cornear() = 0;
};

class Torso
{
    virtual void hablar() = 0;
};

// clase concreta
class Pegaso: Caballo, Alas
{
    string color;
public:
    Pegaso(string color_): color(color_) {}
    void show()
    {
        cout << color << endl;
    }
    void parar(){}
    void volar(){}
};

// clase concreta
class Unicornio: Caballo, Cuerno
{
    void parar(){}
    void cornear(){}
public:
    Unicornio():Caballo(){} //super
    void show()
    {
        cout << color << endl;
    }

};

// clase concreta
class Centauro: Caballo, Torso
{
    string color;
    void parar(){}
    void hablar(){}
public:
    Centauro(string color_): color(color_) {}
    void show()
    {
        cout << color << endl;
    }
};

// clase principal
class MundoMitologico
{
public:
    Pegaso *p = new Pegaso("negro"); 
    Unicornio *u = new Unicornio(); 
    Centauro *c = new Centauro("cafe");
};

int main()
{   
    // declaración en el heap
    MundoMitologico *m = new MundoMitologico();
    m->p->show(); 
    m->u->show();

    // declaración en el stack
    Centauro c2("cafe");
    c2.show(); 
}
